File: clipsync-win/ClipSync/TransferLog.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using System.Threading.Channels;
using ClipSync.Net;

namespace ClipSync;

/// Append-only log of clipboard items the app has sent or received. Used
/// by the test tools to verify that a fuzzed local copy actually crossed
/// the wire. Lines match the mac format so the two logs are diffable:
/// `<iso8601-utc> <SEND|RECV> <bytes> <sha8> "<hint>"`
public static class TransferLog
{
    private static readonly Channel<string> _channel =
        Channel.CreateUnbounded<string>(new UnboundedChannelOptions { SingleReader = true });
    private static readonly string _path = Init();

    public static void Send(ClipboardItem item) => Write("SEND", item);
    public static void Recv(ClipboardItem item) => Write("RECV", item);

    private static string Init()
    {
        var dir = Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
            "ClipSync");
        Directory.CreateDirectory(dir);
        var path = Path.Combine(dir, "transfers.log");
        _ = Task.Run(() => Pump(path));
        return path;
    }

    private static void Write(string direction, ClipboardItem item)
    {
        ulong total = 0;
        foreach (var f in item.Formats) total += f.Size;
        var hash = item.CanonicalHash();
        var sha8 = BitConverter.ToString(hash, 0, 4).Replace("-", "").ToLowerInvariant();
        var hint = Escape(item.Hint ?? "");
        var line = $"{DateTime.UtcNow.ToString("yyyy-MM-ddTHH:mm:ss.fffZ")} {direction} {total} {sha8} \"{hint}\"\n";
        _channel.Writer.TryWrite(line);
    }

    private static async Task Pump(string path)
    {
        await foreach (var line in _channel.Reader.ReadAllAsync())
        {
            try { await File.AppendAllTextAsync(path, line, Encoding.UTF8); } catch { }
        }
    }

    private static string Escape(string s)
    {
        var sb = new StringBuilder(s.Length);
        foreach (var ch in s)
        {
            switch (ch)
            {
                case '\\': sb.Append("\\\\"); break;
                case '"':  sb.Append("\\\""); break;
                case '\n': sb.Append("\\n"); break;
                case '\r': sb.Append("\\r"); break;
                case '\t': sb.Append("\\t"); break;
                default:   sb.Append(ch); break;
            }
        }
        return sb.ToString();
    }
}

File: clipsync-mac/Sources/ClipSync/TransferLog.swift
import Foundation

/// Append-only log of clipboard items the app has sent or received. Used
/// by the test tools to verify that a fuzzed local copy actually crossed
/// the wire. Lines are one per event, intended to be diffable against the
/// peer's log.
///
/// Format: `<iso8601-utc> <SEND|RECV> <bytes> <sha8> "<hint>"`
enum TransferLog {
    private static let queue = DispatcherQueue()

    static let path: URL = {
        let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("ClipSync", isDirectory: true)
        try? FileManager.default.createDirectory(at: base, withIntermediateDirectories: true)
        return base.appendingPathComponent("transfers.log")
    }()

    static func send(_ item: ClipboardItem) { write("SEND", item) }
    static func recv(_ item: ClipboardItem) { write("RECV", item) }

    private static func write(_ direction: String, _ item: ClipboardItem) {
        let total = item.formats.reduce(0) { $0 + Int($1.size) }
        let sha8 = item.canonicalHash.prefix(4).map { String(format: "%02x", $0) }.joined()
        let hint = escape(item.hint ?? "")
        let line = "\(isoNow()) \(direction) \(total) \(sha8) \"\(hint)\"\n"
        queue.run {
            if let data = line.data(using: .utf8) {
                if let fh = try? FileHandle(forWritingTo: path) {
                    fh.seekToEndOfFile()
                    fh.write(data)
                    try? fh.close()
                } else {
                    try? data.write(to: path)
                }
            }
        }
    }

    private static let isoFormatter: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return f
    }()

    private static func isoNow() -> String { isoFormatter.string(from: Date()) }

    private static func escape(_ s: String) -> String {
        var out = ""
        for ch in s {
            switch ch {
            case "\\": out += "\\\\"
            case "\"": out += "\\\""
            case "\n": out += "\\n"
            case "\r": out += "\\r"
            case "\t": out += "\\t"
            default: out.append(ch)
            }
        }
        return out
    }
}

/// Tiny serial queue wrapper so log writes from multiple threads don't
/// interleave on a single line.
private final class DispatcherQueue {
    private let q = DispatchQueue(label: "clipsync.transferlog")
    func run(_ body: @escaping () -> Void) { q.async(execute: body) }
}
